packages/core/src/mesh/ModelMesh.ts

```diff
@@ -669,44 +638,37 @@ export class ModelMesh extends Mesh {
   }
 
   /**
-   * Upload Mesh Data to GPU.
-   * @param noLongerReadable - Whether to read data later. If true, you'll never read data anymore (free memory cache)
+   * Upload data to GPU set by `setPositions()`, `setNormals()`, `setColors()`, `setBoneWeights()`, `setBoneIndices()`, `setTangents()`, `setUVs()`, `setIndices()` methods.
+   * This method will be auto generate vertex element and vertex buffer binding if needed.
+   *
+   * @param releaseData - Whether to release the data cache, release data can reduce memory usage.
    */
-  uploadData(noLongerReadable: boolean): void {
-    // Update vertex elements
+  uploadData(releaseData: boolean): void {
     this._updateVertexElements();
 
-    // Vertex count change
-    const vertexBuffer = this._vertexBufferBindings[0]?._buffer;
-    if (this._vertexCountChanged) {
-      this._vertexBufferUpdateFlag = VertexChangedFlags.All;
-      vertexBuffer?.destroy();
-
-      const elementCount = this._bufferStrides[0] / 4;
-      const vertexFloatCount = elementCount * this.vertexCount;
-      const vertices = new Float32Array(vertexFloatCount);
-      this._verticesFloat32 = vertices;
-      this._verticesUint8 = new Uint8Array(vertices.buffer);
-      this._updateVertices(vertices);
-
-      const bufferUsage = noLongerReadable ? BufferUsage.Static : BufferUsage.Dynamic;
-      const newVertexBuffer = new Buffer(this._engine, BufferBindFlag.VertexBuffer, vertices, bufferUsage);
-
-      this._setVertexBufferBinding(0, new VertexBufferBinding(newVertexBuffer, elementCount * 4));
-      this._vertexCountChanged = false;
-    } else {
-      if (this._vertexBufferUpdateFlag & VertexChangedFlags.All) {
-        let vertices = this._verticesFloat32;
-        if (!vertices) {
-          const elementCount = this._bufferStrides[0] / 4;
-          const vertexFloatCount = elementCount * this.vertexCount;
-          this._verticesFloat32 = vertices = new Float32Array(vertexFloatCount);
+    // If releaseData is false, we shouldn't update buffer data version
+    releaseData || (this._advancedDataSyncToBuffer = true);
+
+    // Update internal vertex buffer if needed
+    this._updateInternalVertexBuffer(releaseData);
+
+    // Update advanced vertex data to buffer
+    if (this._advancedDataUpdateFlag & VertexElementFlags.All) {
+      this._updateAdvancedVertices();
+
+      const vertexBufferInfos = this._vertexBufferInfos;
+      const vertexBufferBindings = this._vertexBufferBindings;
+      for (let i = 0, n = vertexBufferBindings.length; i < n; i++) {
```

Comment:
Unused buffers are not destroyed.
```
    modelMesh.setVertexBufferBinding(posBuffer, 12, 0);
    modelMesh.setVertexBufferBinding(norBuffer, 12, 2);
    modelMesh.setVertexBufferBinding(uvBuffer, 8, 3);
    modelMesh.setNormals(null);
    modelMesh.uploadData(false);
    console.log(buffers, modelMesh._vertexBufferBindings);
```

Reply:
Custom buffer shouldn't destroy, internal buffer maybe!